In [8]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [9]:
client_data = pd.read_csv('client_data (1).csv')
print(client_data.head())

                                 id                     channel_sales  \
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1  d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2  764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3  bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4  149d57cf92fc41cf94415803a877cb4b                           MISSING   

   cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0         0         54946                0  2013-06-15  2016-06-15   
1      4660             0                0  2009-08-21  2016-08-30   
2       544             0                0  2010-04-16  2016-04-16   
3      1584             0                0  2010-03-30  2016-03-30   
4      4425             0              526  2010-01-13  2016-03-07   

  date_modif_prod date_renewal  forecast_cons_12m  ...  has_gas  imp_cons  \
0      2015-11-01   2015-06-23               0.00  ...        t

In [9]:
prices_data = pd.read_csv('price_data (1).csv')
print(prices_data)

                                      id  price_date  price_off_peak_var  \
0       038af19179925da21a25619c5a24b745  2015-01-01            0.151367   
1       038af19179925da21a25619c5a24b745  2015-02-01            0.151367   
2       038af19179925da21a25619c5a24b745  2015-03-01            0.151367   
3       038af19179925da21a25619c5a24b745  2015-04-01            0.149626   
4       038af19179925da21a25619c5a24b745  2015-05-01            0.149626   
...                                  ...         ...                 ...   
192997  16f51cdc2baa19af0b940ee1b3dd17d5  2015-08-01            0.119916   
192998  16f51cdc2baa19af0b940ee1b3dd17d5  2015-09-01            0.119916   
192999  16f51cdc2baa19af0b940ee1b3dd17d5  2015-10-01            0.119916   
193000  16f51cdc2baa19af0b940ee1b3dd17d5  2015-11-01            0.119916   
193001  16f51cdc2baa19af0b940ee1b3dd17d5  2015-12-01            0.119916   

        price_peak_var  price_mid_peak_var  price_off_peak_fix  \
0             0.00000

In [7]:
df = pd.read_csv('./clean_data_after_eda.csv')
df["date_activ"] = pd.to_datetime(df["date_activ"], format='%Y-%m-%d')
df["date_end"] = pd.to_datetime(df["date_end"], format='%Y-%m-%d')
df["date_modif_prod"] = pd.to_datetime(df["date_modif_prod"], format='%Y-%m-%d')
df["date_renewal"] = pd.to_datetime(df["date_renewal"], format='%Y-%m-%d')

In [11]:
price_df = pd.read_csv('price_data (1).csv')
price_df["price_date"] = pd.to_datetime(price_df["price_date"], format='%Y-%m-%d')

In [11]:
print(client_data.dtypes)

id                                 object
channel_sales                      object
cons_12m                            int64
cons_gas_12m                        int64
cons_last_month                     int64
date_activ                         object
date_end                           object
date_modif_prod                    object
date_renewal                       object
forecast_cons_12m                 float64
forecast_cons_year                  int64
forecast_discount_energy          float64
forecast_meter_rent_12m           float64
forecast_price_energy_off_peak    float64
forecast_price_energy_peak        float64
forecast_price_pow_off_peak       float64
has_gas                            object
imp_cons                          float64
margin_gross_pow_ele              float64
margin_net_pow_ele                float64
nb_prod_act                         int64
net_margin                        float64
num_years_antig                     int64
origin_up                         

In [15]:
print(prices_data.dtypes)

id                     object
price_date             object
price_off_peak_var    float64
price_peak_var        float64
price_mid_peak_var    float64
price_off_peak_fix    float64
price_peak_fix        float64
price_mid_peak_fix    float64
dtype: object


In [13]:
prices_monthly_by_id = price_df.groupby(['id', 'price_date']).agg({'price_off_peak_var': 'mean', 'price_off_peak_fix': 'mean'}).reset_index()
print(prices_monthly_by_id)


                                      id price_date  price_off_peak_var  \
0       0002203ffbb812588b632b9e628cc38d 2015-01-01            0.126098   
1       0002203ffbb812588b632b9e628cc38d 2015-02-01            0.126098   
2       0002203ffbb812588b632b9e628cc38d 2015-03-01            0.128067   
3       0002203ffbb812588b632b9e628cc38d 2015-04-01            0.128067   
4       0002203ffbb812588b632b9e628cc38d 2015-05-01            0.128067   
...                                  ...        ...                 ...   
192997  ffff7fa066f1fb305ae285bb03bf325a 2015-08-01            0.119916   
192998  ffff7fa066f1fb305ae285bb03bf325a 2015-09-01            0.119916   
192999  ffff7fa066f1fb305ae285bb03bf325a 2015-10-01            0.119916   
193000  ffff7fa066f1fb305ae285bb03bf325a 2015-11-01            0.119916   
193001  ffff7fa066f1fb305ae285bb03bf325a 2015-12-01            0.119916   

        price_off_peak_fix  
0                40.565969  
1                40.565969  
2           

In [45]:
jan_prices = prices_monthly_by_id.groupby(['id']).first().reset_index()
Dec_prices = prices_monthly_by_id.groupby(['id']).last().reset_index()

merged_jan_dec_prices = pd.merge(Dec_prices.rename(columns={'price_off_peak_var': 'dec_1', 'price_off_peak_fix': 'dec_2'}), jan_prices.drop(columns='price_date'), on='id')
print(merged_jan_dec_prices)

                                     id price_date     dec_1      dec_2  \
0      0002203ffbb812588b632b9e628cc38d 2015-12-01  0.119906  40.728885   
1      0004351ebdd665e6ee664792efc4fd13 2015-12-01  0.143943  44.444710   
2      0010bcc39e42b3c2131ed2ce55246e3c 2015-12-01  0.201280  45.944710   
3      0010ee3855fdea87602a5b7aba8e42de 2015-12-01  0.113068  40.728885   
4      00114d74e963e47177db89bc70108537 2015-12-01  0.145440  44.266930   
...                                 ...        ...       ...        ...   
16091  ffef185810e44254c3a4c6395e6b4d8a 2015-12-01  0.112488  40.728885   
16092  fffac626da707b1b5ab11e8431a4d0a2 2015-12-01  0.145047  44.444710   
16093  fffc0cacd305dd51f316424bbb08d1bd 2015-12-01  0.151399  41.228885   
16094  fffe4f5646aa39c7f97f95ae2679ce64 2015-12-01  0.118175  40.728885   
16095  ffff7fa066f1fb305ae285bb03bf325a 2015-12-01  0.119916  40.728885   

       price_off_peak_var  price_off_peak_fix  
0                0.126098           40.565969  
1  

In [49]:
merged_jan_dec_prices['diff_var_off_peak_dec_jan'] = merged_jan_dec_prices['dec_1'] - merged_jan_dec_prices['price_off_peak_var']
merged_jan_dec_prices['diff_fixed_off_peak_dec_jan'] = merged_jan_dec_prices['dec_2'] - merged_jan_dec_prices['price_off_peak_fix']
merged_jan_dec_prices = merged_jan_dec_prices[['id', 'diff_var_off_peak_dec_jan', 'diff_fixed_off_peak_dec_jan']]
merged_jan_dec_prices.head()

id  diff_var_off_peak_dec_jan  \
0  0002203ffbb812588b632b9e628cc38d                  -0.006192   
1  0004351ebdd665e6ee664792efc4fd13                  -0.004104   
2  0010bcc39e42b3c2131ed2ce55246e3c                   0.050443   
3  0010ee3855fdea87602a5b7aba8e42de                  -0.010018   
4  00114d74e963e47177db89bc70108537                  -0.003994   

   diff_fixed_off_peak_dec_jan  
0                     0.162916  
1                     0.177779  
2                     1.500000  
3                     0.162916  
4                    -0.000001

In [51]:
pd.merge(df,merged_jan_dec_prices, on='id')

id                     channel_sales  \
0      24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1      d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2      764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3      bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4      149d57cf92fc41cf94415803a877cb4b                           MISSING   
...                                 ...                               ...   
14601  18463073fb097fc0ac5d3e040f356987  foosdfpfkusacimwkcsosbicdxkicaua   
14602  d0a6f71671571ed83b2645d23af6de00  foosdfpfkusacimwkcsosbicdxkicaua   
14603  10e6828ddd62cbcf687cb74928c4c2d2  foosdfpfkusacimwkcsosbicdxkicaua   
14604  1cf20fd6206d7678d5bcafd28c53b4db  foosdfpfkusacimwkcsosbicdxkicaua   
14605  563dde550fd624d7352f3de77c0cdfcd                           MISSING   

       cons_12m  cons_gas_12m  cons_last_month date_activ   date_end  \
0             0         54946                0 2013-06-15 2016-06-15   
1          4660             0                0 2009-08-21 2016-08-30   
2           544             0                0 2010-04-16 2016-04-16   
3          1584             0                0 2010-03-30 2016-03-30   
4          4425             0              526 2010-01-13 2016-03-07   
...         ...           ...              ...        ...        ...   
14601     32270         47940                0 2012-05-24 2016-05-08   
14602      7223             0              181 2012-08-27 2016-08-27   
14603      1844             0              179 2012-02-08 2016-02-07   
14604       131             0                0 2012-08-30 2016-08-30   
14605      8730             0                0 2009-12-18 2016-12-17   

      date_modif_prod date_renewal  forecast_cons_12m  ...  \
0          2015-11-01   2015-06-23               0.00  ...   
1          2009-08-21   2015-08-31             189.95  ...   
2          2010-04-16   2015-04-17              47.96  ...   
3          2010-03-30   2015-03-31             240.04  ...   
4          2010-01-13   2015-03-09             445.75  ...   
...               ...          ...                ...  ...   
14601      2015-05-08   2014-05-26            4648.01  ...   
14602      2012-08-27   2015-08-28             631.69  ...   
14603      2012-02-08   2015-02-09             190.39  ...   
14604      2012-08-30   2015-08-31              19.34  ...   
14605      2009-12-18   2015-12-21             762.41  ...   

       var_6m_price_mid_peak_var  var_6m_price_off_peak_fix  \
0                   9.084737e-04                   2.086294   
1                   0.000000e+00                   0.009482   
2                   0.000000e+00                   0.000000   
3                   0.000000e+00                   0.000000   
4                   4.860000e-10                   0.000000   
...                          ...                        ...   
14601               0.000000e+00                   0.000000   
14602               1.802667e-08                   0.014939   
14603               4.860000e-10                   0.000000   
14604               0.000000e+00                   0.009482   
14605               0.000000e+00                   0.000000   

       var_6m_price_peak_fix  var_6m_price_mid_peak_fix  \
0                  99.530517                  44.235794   
1                   0.000000                   0.000000   
2                   0.000000                   0.000000   
3                   0.000000                   0.000000   
4                   0.000000                   0.000000   
...                      ...                        ...   
14601               0.000000                   0.000000   
14602               0.005682                   0.000299   
14603               0.000000                   0.000000   
14604               0.000000                   0.000000   
14605               0.000000                   0.000000   

       var_6m_price_off_peak  var_6m

In [53]:
# Aggregate average prices per period by company
mean_prices = price_df.groupby(['id']).agg({
    'price_off_peak_var': 'mean', 
    'price_peak_var': 'mean', 
    'price_mid_peak_var': 'mean',
    'price_off_peak_fix': 'mean',
    'price_peak_fix': 'mean',
    'price_mid_peak_fix': 'mean'    
}).reset_index()
# Calculate the mean difference between consecutive periods
mean_prices['off_peak_peak_var_mean_diff'] = mean_prices['price_off_peak_var'] - mean_prices['price_peak_var']
mean_prices['peak_mid_peak_var_mean_diff'] = mean_prices['price_peak_var'] - mean_prices['price_mid_peak_var']
mean_prices['off_peak_mid_peak_var_mean_diff'] = mean_prices['price_off_peak_var'] - mean_prices['price_mid_peak_var']
mean_prices['off_peak_peak_fix_mean_diff'] = mean_prices['price_off_peak_fix'] - mean_prices['price_peak_fix']
mean_prices['peak_mid_peak_fix_mean_diff'] = mean_prices['price_peak_fix'] - mean_prices['price_mid_peak_fix']
mean_prices['off_peak_mid_peak_fix_mean_diff'] = mean_prices['price_off_peak_fix'] - mean_prices['price_mid_peak_fix']
columns = [
    'id', 
    'off_peak_peak_var_mean_diff',
    'peak_mid_peak_var_mean_diff', 
    'off_peak_mid_peak_var_mean_diff',
    'off_peak_peak_fix_mean_diff', 
    'peak_mid_peak_fix_mean_diff', 
    'off_peak_mid_peak_fix_mean_diff'
]
df = pd.merge(df, mean_prices[columns], on='id')
df.head()

id                     channel_sales  \
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1  d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2  764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3  bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4  149d57cf92fc41cf94415803a877cb4b                           MISSING   

   cons_12m  cons_gas_12m  cons_last_month date_activ   date_end  \
0         0         54946                0 2013-06-15 2016-06-15   
1      4660             0                0 2009-08-21 2016-08-30   
2       544             0                0 2010-04-16 2016-04-16   
3      1584             0                0 2010-03-30 2016-03-30   
4      4425             0              526 2010-01-13 2016-03-07   

  date_modif_prod date_renewal  forecast_cons_12m  ...  var_6m_price_off_peak  \
0      2015-11-01   2015-06-23               0.00  ...               2.086425   
1      2009-08-21   2015-08-31             189.95  ...               0.009485   
2      2010-04-16   2015-04-17              47.96  ...               0.000004   
3      2010-03-30   2015-03-31             240.04  ...               0.000003   
4      2010-01-13   2015-03-09             445.75  ...               0.000011   

   var_6m_price_peak  var_6m_price_mid_peak  churn  \
0       9.953056e+01           4.423670e+01      1   
1       1.217891e-03           0.000000e+00      0   
2       9.450150e-08           0.000000e+00      0   
3       0.000000e+00           0.000000e+00      0   
4       2.896760e-06           4.860000e-10      0   

   off_peak_peak_var_mean_diff  peak_mid_peak_var_mean_diff  \
0                     0.024038                     0.034219   
1                     0.142485                     0.007124   
2                     0.082090                     0.088421   
3                     0.151210                     0.000000   
4                     0.020536                     0.030773   

  off_peak_mid_peak_var_mean_diff  off_peak_peak_fix_mean_diff  \
0                        0.058257                    18.590255   
1                        0.149609                    44.311375   
2                        0.170512                    44.385450   
3                        0.151210                    44.400265   
4                        0.051309                    16.275263   

   peak_mid_peak_fix_mean_diff  off_peak_mid_peak_fix_mean_diff  
0                     7.450670                        26.040925  
1                     0.000000                        44.311375  
2                     0.000000                        44.385450  
3                     0.000000                        44.400265  
4                     8.137629                        24.412893  

[5 rows x 50 columns]

In [97]:
total_churned = (client_data['churn'] == 1).sum()
print(total_churned)
total_values = len(client_data)
print(total_values)
total_churned_percent = total_churned / total_values
print(total_churned_percent)

1419
14606
0.09715185540188963


less than 10% of all clients churned over the next 3 months

The total nmber of clients that churned in the last 3 months is 1419.

In [34]:
total_churned_gas_and_electricity = ((client_data['churn'] == 1) & (client_data['has_gas'] == 't')).sum()
print(total_churned_gas_and_electricity)

217


In [38]:
percent_of_churned_for_gas_and_electricity = total_churned_gas_and_electricity / total_churned
print(percent_of_churned_for_gas_and_electricity)

0.1529245947850599


from those who have both gas and electricity 15% of them churned within the next 3 months

In [10]:
clients_data_max_usage_over_last_12_months_50 = client_data.sort_values(by='imp_cons', ascending = False).head(50)
print(clients_data_max_usage_over_last_12_months_50.head())

                                     id                     channel_sales  \
10219  836bd604bb49e6c3e717403a466a0407                           MISSING   
295    a33668b7d0d8d86f38ad983c0cad6095  foosdfpfkusacimwkcsosbicdxkicaua   
7318   96974bd838f708c58e1cbebaab4a052b                           MISSING   
12927  e5962682808930babc6561b0b6e5050a  lmkebamcaaclubfxadlmueccxoimlema   
7794   a3a739686fbd5ba8b4a21ec835507b6d  foosdfpfkusacimwkcsosbicdxkicaua   

       cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
10219    222615             0           175375  2008-02-12  2016-02-12   
295       63969          7488            63969  2013-06-13  2017-01-06   
7318      92477             0            59460  2009-10-28  2016-10-28   
12927    156003             0            70180  2010-06-21  2016-02-15   
7794    1086232             0            82136  2012-02-01  2016-02-01   

      date_modif_prod date_renewal  forecast_cons_12m  ...  has_gas  imp_cons  \
10219      

In [159]:
(clients_data_max_usage_over_last_12_months_50['churn'] == 1).sum() / (len(clients_data_max_usage_over_last_12_months_50)) 

0.12

from the top 50 highest current paid consumptions from all clients only 6 of the 50 churned over the next 3 months
12% of the highest current paid consumers churned over the next 3 months and 10% churned overall 